### PR TITLE
[Merged by Bors] - Parallelize forward kinematics animation systems

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Speed up animation by leveraging all threads in `ComputeTaskPool`.

## Solution
This PR parallelizes animation sampling across all threads. 

To ensure that this is safely done, all animation is predicated with an ancestor query to ensure that there is no conflicting `AnimationPlayer` above each animated hierarchy that may cause this to alias.

Unlike the RFC, this does not add support for reflect based "animate anything", but only extends the existing `AnimationPlayer` to support high numbers of animated characters on screen at once.

## Performance
This cuts `many_foxes`'s frame time on my machine by a full millisecond, from 7.49ms to 6.5ms. (yellow is this PR, red is main).
image

---

## Changelog
Changed: Animation sampling now runs fully multi-threaded using threads from `ComputeTaskPool`.
Changed: `AnimationPlayer` that are on a child or descendant of another entity with another player will no longer be run.

### CODE REVIEW

**Comment by mockersf** [comment]
this PR decreased performances for me in WASM, from 48fps to 42

**Comment by james7132** [comment]
Makes plenty of sense, this change adds a sizable amount of extra actual work that only makes sense in a system with actual parallelism.

Could cfg block this for now, but the real solution may involve figuring out WASM parallelism.

**Comment by mockersf** [comment]
single threaded is a valid scenario not just for wasm, #6690

**Comment by james7132** [comment]
Added a quick and dirty entity-path caching solution. This shaved off yet another 1ms off of my local runtime for `many_foxes`.

image

**Comment by superdump** [comment]
The gains are awesome!

**Comment by james7132** [comment]
Switched up the strategy. Instead of using `BoneBindings` as an intermediate deduplication step, it predicates any animation with an ancestor query to ensure there is no `AnimationPlayer` in it's ancestors to ensure that there's no other mutation happening. This should remove the large overhead that you were seeing on WASM/single threaded builds.

This means we cannot support cases where there's an animation player on a child/descendant of another animation player, but I think that's a fairly niche edge case we can safely ignore.

This shaved yet another 1ms off of the frame time of `many_foxes` on my machine. We're above 200FPS with Tracy enabled now!

image

File: crates/bevy_ecs/src/storage/sparse_set.rs [hunk 1]
```diff
@@ -447,6 +447,12 @@ impl<I: SparseSetIndex, V> SparseSet<I, V> {
         })
     }
 
+    pub fn clear(&mut self) {
+        self.dense.clear();
+        self.indices.clear();
+        self.sparse.clear();
+    }
+
```
Comment (mockersf) [inline crates/bevy_ecs/src/storage/sparse_set.rs:455]: this seems unrelated?

File: crates/bevy_animation/Cargo.toml [hunk 2]
```diff
@@ -19,4 +19,8 @@ bevy_time = { path = "../bevy_time", version = "0.9.0" }
 bevy_utils = { path = "../bevy_utils", version = "0.9.0" }
 bevy_ecs = { path = "../bevy_ecs", version = "0.9.0" }
 bevy_transform = { path = "../bevy_transform", version = "0.9.0" }
+bevy_tasks = { path = "../bevy_tasks", version = "0.9.0" }
 bevy_hierarchy = { path = "../bevy_hierarchy", version = "0.9.0" }
+
+smallvec = "1.10"
+thread_local = "1.1"
```
Comment (mockersf) [inline crates/bevy_animation/Cargo.toml:26]:
```suggestion
thread_local = "1.1"
```

**Comment by mockersf** [comment]
> This means we cannot support cases where there's an animation player on a child/descendant of another animation player, but I think that's a fairly niche edge case we can safely ignore.

I agree, that's a case that can be ignored in the default animation player. Users can roll out their own if they want to support that, or just merge their animations as one so that they can play from the same node.

Alternatively, we could mark those as not doable in parallel and do them after one at a time with a perf cost.

Looks good except for two small nits, and perf are good for me 👍

File: crates/bevy_ecs/src/storage/sparse_set.rs (hunk 1 above)
Comment (james7132) [reply crates/bevy_ecs/src/storage/sparse_set.rs:455]: Oh that was a remnant from the old approach. I'll remove it.

File: crates/bevy_animation/Cargo.toml (hunk 2 above)
Comment (james7132) [reply crates/bevy_animation/Cargo.toml]: These actually aren't needed anymore. Removed them.

**Review by mockersf — APPROVED** [review]

File: crates/bevy_animation/src/lib.rs [hunk 3]
```diff
@@ -63,17 +64,34 @@ pub struct EntityPath {
...
 pub struct AnimationClip {
-    curves: HashMap<EntityPath, Vec<VariableCurve>>,
+    curves: Vec<Vec<VariableCurve>>,
+    paths: HashMap<EntityPath, usize>,
     duration: f32,
 }
 
 impl AnimationClip {
     #[inline]
     /// Hashmap of the [`VariableCurve`]s per [`EntityPath`].
```
Comment (hymm) [inline crates/bevy_animation/src/lib.rs:74]:
```suggestion
    /// Vec<Vec<_>> of the [`VariableCurve`]s per [`EntityPath`].
```
Not sure that's the correct way of phrasing that

**Review by hymm — APPROVED** [review]
This is pretty complicated code now, so I would eventually like some tests for this code, but that doesn't need to be in this PR.

I gain about 2ms on my machine for many_foxes.

One small nit.

File: crates/bevy_animation/src/lib.rs (hunk 3 above)
Comment (hymm) [reply crates/bevy_animation/src/lib.rs:74]: maybe just remove the return type from the commnet.

Comment (james7132) [reply crates/bevy_animation/src/lib.rs]: Done.

**Review by cart — APPROVED** [review]
Very nice wins! Fun seeing ideas from this RFC in the wild: https://github.com/bevyengine/rfcs/pull/51

I'm also getting a ~2ms win on many_foxes
